core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PushFilterIntoValues.java

```diff
@@ -0,0 +1,165 @@
+import static java.util.Objects.requireNonNull;
+
+public class PushFilterIntoValues
+        implements Rule<FilterNode>
+{
+    private static final Capture<ValuesNode> VALUES = newCapture();
+
+    private static final Pattern<FilterNode> PATTERN = filter()
+            .with(source().matching(values()
+                    .matching(PushFilterIntoValues::isSupportedValues)
+                    .capturedAs(VALUES)));
+
+    private final PlannerContext plannerContext;
+
+    public PushFilterIntoValues(PlannerContext plannerContext)
+    {
+        this.plannerContext = requireNonNull(plannerContext, "plannerContext is null");
+    }
+
+    @Override
+    public boolean isEnabled(Session session)
+    {
+        return isPushFilterIntoValues(session);
+    }
+
+    @Override
+    public Pattern<FilterNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(FilterNode node, Captures captures, Context context)
+    {
+        ValuesNode valuesNode = captures.get(VALUES);
+
+        Expression predicate = node.getPredicate();
+        if (!isDeterministic(predicate)) {
+            return Result.empty();
+        }
+
+        if (valuesNode.getRows().orElseThrow().isEmpty()) {
+            // empty values node, filter is redundant
+            return Result.ofPlanNode(valuesNode);
+        }
+
+        if (valuesNode.getRows().orElseThrow().stream()
+                .anyMatch(row -> !isDeterministic(row))) {
+            return Result.empty();
+        }
+
+        Set<Symbol> valuesExpressionsSymbols = valuesNode.getRows().orElseThrow().stream()
+                .map(SymbolsExtractor::extractUnique)
+                .flatMap(Set::stream)
+                .collect(toImmutableSet());
+        if (!valuesExpressionsSymbols.isEmpty()) {
+            // unresolved correlation
+            return Result.empty();
+        }
+
+        Set<Symbol> valuesOutputSymbols = ImmutableSet.copyOf(valuesNode.getOutputSymbols());
+        Set<Symbol> predicateSymbols = extractUnique(predicate);
+        if (!valuesOutputSymbols.containsAll(predicateSymbols)) {
+            // filter is correlated
+            return Result.empty();
+        }
+
+        ImmutableList.Builder<Expression> filteredRows = ImmutableList.builder();
+        boolean optimized = false;
+        boolean keepFilter = false;
+        for (Expression expression : valuesNode.getRows().orElseThrow()) {
+            Row row = (Row) expression;
+            ImmutableMap.Builder<Symbol, Expression> mapping = ImmutableMap.builder();
+            for (int i = 0; i < valuesNode.getOutputSymbols().size(); i++) {
+                mapping.put(valuesNode.getOutputSymbols().get(i), row.items().get(i));
+            }
+            Expression rewrittenPredicate = inlineSymbols(mapping.buildOrThrow(), predicate);
+            Expression optimizedPredicate = new IrExpressionInterpreter(rewrittenPredicate, plannerContext, context.getSession()).optimize();
```

Comment:
Can this be expensive when VALUES has eg 1000 rows?

Can we have even bigger VALUES, eg as a result of inlining or eager evaluaiton of something else?

Reply:
I am not aware of any mechanism that creates bigger VALUES, in terms of the number of rows, than those originally present in the query.

Reply:
i am not aware either, but it's possible to imagine such a mechanism to exist in the future.
https://github.com/trinodb/trino/issues/684 was created in context of some discussions we had. for example, it was considered that we eagerly read and inline small tables "during planning" to inform the planner better and produce optimal plan overall. That wouldn't be conceptually much different than what we do in `ExtractSpatialJoins`.